rafx-shader-processor/src/parse_declarations.rs

```diff
@@ -555,15 +572,94 @@ fn try_parse_binding(code: &[char]) -> Result<Option<ParseBindingResult>, String
     }))
 }
 
-fn try_parse_const(code: &[char]) -> Result<Option<()>, String> {
+fn try_parse_const(code: &[char]) -> Result<Option<ParseConstantResult>, String> {
     let mut position = 0;
 
-    // Consume the layout keyword. If it's missing, assume this isn't a binding and return None
+    //
+    // See if it starts with const. If not, assume this isn't a constant and return None
+    //
     if crate::parse_source::try_consume_literal(code, &mut position, "const").is_none() {
         return Ok(None);
     }
 
-    Ok(Some(()))
+    crate::parse_source::skip_whitespace(code, &mut position);
+    let type_name =
+        crate::parse_source::try_consume_identifier(code, &mut position).ok_or(format!(
+            "Expected type name while parsing constant:\n{}",
+            crate::parse_source::characters_to_string(&code)
+        ))?;
+
+    crate::parse_source::skip_whitespace(code, &mut position);
+    let instance_name =
+        crate::parse_source::try_consume_identifier(code, &mut position).ok_or(format!(
+            "Expected instance name while parsing constant:\n{}",
+            crate::parse_source::characters_to_string(&code)
+        ))?;
+
+    crate::parse_source::skip_whitespace(code, &mut position);
+    crate::parse_source::try_consume_literal(code, &mut position, "=").ok_or(format!(
+        "Expected = while parsing constant:\n{}",
+        crate::parse_source::characters_to_string(&code)
+    ))?;
+
+    crate::parse_source::skip_whitespace(code, &mut position);
+    if crate::parse_source::is_string_at_position(code, position, "vec3")
+        || crate::parse_source::is_string_at_position(code, position, "vec4")
+        || crate::parse_source::is_string_at_position(code, position, "mat3")
+        || crate::parse_source::is_string_at_position(code, position, "mat4")
+    {
+        // TODO(dvd): We don't support mat3, mat4, vec3, vec4.
+        return Ok(Some(ParseConstantResult {
+            type_name,
+            instance_name,
+            value: None,
+        }));
+    }
+
+    // 1 or more values with + / - * between them.
+    let mut values = Vec::default();
+    loop {
+        values.push(
+            crate::parse_source::try_consume_constant_value(code, &mut position).ok_or(format!(
+                "Expected value while parsing constant:\n{}",
+                crate::parse_source::characters_to_string(&code)
+            ))?,
+        );
+
+        if crate::parse_source::try_consume_literal(code, &mut position, ";").is_some() {
+            break;
+        } else {
+            fn push_operator(
+                values: &mut Vec<String>,
+                operator: &str,
+            ) {
+                values.push(format!(" {} ", operator));
+            }
+
+            crate::parse_source::skip_whitespace(code, &mut position);
```

Comment:
This parsing is actually pretty problematic, and given what it accomplishes, I think it would be cleaner to just copy the entire string between the `=` and the `;`. The issue is that this definition would ban trivial things like using parentheses. 

I think if this was rewritten to just return the string between `=` and `;`, then later code in the pipeline can accommodate that accordingly -- like in this PR, where an array length is simply not allowed to be anything other than a trivial value.

Or, if possible, maybe we could _remove_ the need to actually parse the `usize` for array lengths now that we can embed the `const` itself into the generated Rust code. I'm not sure if anything in the shader processor depends on knowing the actual array length (as a number) vs knowing a symbol (like the name of a `const` identifier).

Reply:
we may need to know array lengths of certain things like texture arrays so we can flag the binding as being an array of N textures.

Reply:
I think the way this is done now is fine. It's reasonable behavior, not complicated, and we can always build on it later.